src/schema-builder/RdbmsSchemaBuilder.ts

```diff
@@ -407,24 +406,13 @@ export class RdbmsSchemaBuilder implements SchemaBuilder {
      */
     protected async dropColumnReferencedIndices(tableName: string, columnName: string): Promise<void> {
 
-        const allIndexMetadatas = this.connection.entityMetadatas.reduce(
-            (all, metadata) => all.concat(metadata.indices),
-            [] as IndexMetadata[]
-        );
-
         const tableSchema = this.tableSchemas.find(table => table.name === tableName);
         if (!tableSchema)
             return;
 
         // find depend indices to drop them
-        const dependIndices = allIndexMetadatas.filter(indexMetadata => {
```

Comment:
can you please explain this removal?

Reply:
Yes sorry, it was in the commit message. I am putting it here as well

We change the logic for finding the indices on dropped columns. Instead of searching in the metadata we just look in the retrieved schema.

We change this logic for 2 reasons:

1. If the user of TORM drops a column then it is his responsibility to maintain the rest of the schema, removing all references of this column from everywhere. TORM has a validator for that thus it catches such errors on startup. In any case the previous code was leaving the dropped index inside the metadata but was removing it from DB which is bad.

2. If the index is removed from the metadata, then all checks where failing since the allIndexMetadata filtering was not returning anything

Reply:
Im sorry but I didnt not understand both reasons.

> previous code was leaving the dropped index inside the metadata but was removing it from DB which is bad.

what does it mean? How it can leave dropped index inside metadata? If user removes index from entity it will be dropped from the database. If user added index to entity it will be created.


> If the user of TORM drops a column then it is his responsibility to maintain the rest of the schema, removing all references of this column from everywhere.

We should help him everywhere we really can to maintain his schema because maintaining schema is a pain and we are trying to abstract people from this process.

In our case I think we can remove indices if he removed them from entity and change or add them when he does it, right?

Reply:
Hi again, please let me clarify. 

Let's say the user wants to remove a column. In this case it is normal to assume that he will also remove all references to this column (indices, foreign keys etc) from the rest of the metadata. If not there should be proper validations. This works perfectly, the validations are there.

But the code I removed implies that although the column is removed, the reference of the column remains inside the metadata and it searches for the such a reference in the indices (of the metadata) when it should be instead searching for such a reference in the indices of the db schema (which is not removed yet). 

Since the reference of the column is not inside the metadata - the user has also removed it (otherwise the schema validations that run early would have caught such bad configuration), this search fails so and the system can not find what indices needs to be removed.

Regarding the second point, you do this already. You have the validations and it is perfect.

Reply:
There is only one hole in this system that I see and it looked like this.
The user has a table where a column is referenced by an index (as an example).

The user removes the column from the metadata at runtime (removing it from the metadata array) and forgets to remove the reference in the index (or just remove the entire index). This breaks at runtime, but I do not consider this a proper way to do it. We can catch this if we run the validations early every time the schema is updating, not just on create connection. This is bullet proof.

Or you can change the entire metadata structure and instead of providing direct reference to the Columns array, wrap it in an ColumnList class, and implement add and remove methods.
At remove you can also remove all references of this column from other schema objects.

That's what i meant by assisting the user on schema changes but I believe that it's the wrong thing to do. 

Hope this clears things a bit

Thanks